src/ws/cockpitauth.c

```diff
@@ -128,6 +130,104 @@ cockpit_authenticated_free (gpointer data)
   g_free (authenticated);
 }
 
+typedef struct {
+  CockpitAuthPipe *auth_pipe;
+  gchar *id;
+  gchar *response_data;
+  GSimpleAsyncResult *pending_result;
+
+  gint refs;
+  gpointer tag;
+  gpointer user_data;
+  GDestroyNotify destroy_func;
+} AuthData;
+
+static void
+auth_data_free (gpointer data)
+{
+  AuthData *ad = data;
+
+  if (ad->auth_pipe)
+    g_object_unref (ad->auth_pipe);
+
+  if (ad->destroy_func && ad->user_data)
+    ad->destroy_func (ad->user_data);
+
+  g_free (ad->id);
+  g_free (ad->response_data);
+
+  g_return_if_fail (ad->pending_result == NULL);
+  g_free (ad);
+}
+
+static void
+auth_data_add_pending_result (AuthData *data,
+                              GSimpleAsyncResult *result)
+{
+  g_return_if_fail (data->pending_result == NULL);
+  data->pending_result = g_object_ref (result);
+}
+
+static void
+auth_data_complete_result (AuthData *data,
+                           GError *error)
+{
+  if (data->pending_result)
+    {
+      if (error)
+        g_simple_async_result_set_from_error (data->pending_result, error);
+
+      g_simple_async_result_complete_in_idle (data->pending_result);
+
+      g_object_unref (data->pending_result);
+      data->pending_result = NULL;
+    }
+  else if (error)
+    {
+      g_message ("Dropped authentication error: %s no pending request to respond to", error->message);
+    }
+  else
+    {
+      g_message ("Dropped authentication result, no pending request to respond to");
+    }
+}
+
+static void
+on_auth_pipe_result (CockpitAuthPipe *auth_pipe,
+                     GBytes *message,
+                     GError *error,
+                     gpointer user_data)
+{
+  AuthData *ad = user_data;
+  gsize len;
+
+  if (message && !error)
+    {
+      len = g_bytes_get_size (message);
+      g_return_if_fail (ad->response_data == NULL);
+      ad->response_data = g_strndup (g_bytes_get_data(message, NULL), len);
+    }
+
+  auth_data_complete_result (ad, error);
+}
+
+static AuthData *
+auth_data_ref (AuthData *data)
+{
+  g_return_val_if_fail (data != NULL, NULL);
+  g_atomic_int_inc (&data->refs);
```

Comment:
Does this need to be thread safe? The use of the atomic functions seem to indicate that it does. However we mutate the structure elsewhere, so I would say "no". Should we just use ```++``` here instead?